python/tk_3de4/menu_generation.py

```diff
@@ -0,0 +1,216 @@
+        with open(script_path, "w") as menu_file:
+            menu_file.write("# 3DE4.script.name: {}\n".format(name))
+            menu_file.write("# 3DE4.script.gui:	{}\n".format(parent_menu))
+            menu_file.write("\n".join(script))
+
+    def _add_command_to_menu(self, cmd, parent_menu, favourite=False):
+        """
+        Take a command and add it to the specified menu.
+
+        :param AppCommand cmd: The command to add.
+        :param str parent_menu: The name of the parent menu item.
+        :param bool favourite: Is the command a favourite command.
+        """
+        script = [
+            "import sgtk",
+            "if __name__ == '__main__':",
+            "   engine = sgtk.platform.current_engine()",
+            "   engine.commands[{!r}]['callback']()".format(cmd.name),
+        ]
+        name = cmd.name
+        if favourite:
+            # Dashes come after spaces but before letters in ordering
+            name = "- {}".format(cmd.name)
+        self._add_script_to_menu(name, parent_menu, script)
+
+
+class AppCommand(object):
+    """
+    Wraps around a single command that you get from engine.commands
+    """
+
+    def __init__(self, name, command_dict):
+        """
+        Initialise the class.
+
+        :param str name: The name fo the command.
+        :param dict command_dict: The dictionary defining this command.
+        """
+        self.name = name
+        self.properties = command_dict["properties"]
+        self.callback = command_dict["callback"]
+        self.favourite = self._is_app_favourite()
+
+    def _is_app_favourite(self):
+        """
+        Figure out if this command is specified as a favourite.
+
+        :rtype: bool
+        """
+        if "app" not in self.properties:
+            return False
+        app_instance = self.properties["app"]
+        engine = app_instance.engine
+        app_instance_name = self.get_app_instance_name()
+        for fav in engine.get_setting("menu_favourites"):
+            fav_app_instance_name = fav["app_instance"]
+            menu_name = fav["name"]
+            if app_instance_name == fav_app_instance_name and self.name == menu_name:
+                return True
+        return False
+
+    def get_app_name(self):
+        """
+        Returns the name of the app that this command belongs to.
+
+        :rtype: str
+        """
+        if "app" in self.properties:
+            return self.properties["app"].display_name
+        return "Other Items"
+
+    def get_app_instance_name(self):
+        """
+        Returns the name of the app instance, as defined in the environment.
+        Returns None if not found.
+
+        :rtype: str or Nonetype
+        """
+        if "app" not in self.properties:
+            return None
+
+        app_instance = self.properties["app"]
+        engine = app_instance.engine
+
+        for (app_instance_name, app_instance_obj) in engine.apps.items():
```

Comment:
Remove redundant braces
```suggestion
        for app_instance_name, app_instance_obj in engine.apps.items():
```